// @mongosh/shell-bson/lib/printable-bson.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeBsonStringifiers = void 0;
exports.makePrintableBson = makePrintableBson;
const util_1 = require("util");
const inspectCustom = Symbol.for('nodejs.util.inspect.custom');
function removeNewFromInspectResult(str) {
    return str.replace(/^(\s*)(new )/, '$1');
}
function removeTypedArrayPrefixFromInspectResult(str) {
    return str.replace(/^\s*\S+\s*\(\d+\)\s*/, '');
}
function makeClasslessInspect(cls) {
    const originalInspect = cls.prototype.inspect;
    return function (...args) {
        return removeNewFromInspectResult(originalInspect.apply(this, args));
    };
}
const makeBinaryVectorInspect = (bsonLibrary) => {
    return function (depth, options) {
        const binaryInspect = makeClasslessInspect(bsonLibrary.Binary);
        switch (this.buffer[0]) {
            case bsonLibrary.Binary.VECTOR_TYPE.Int8:
                return `Binary.fromInt8Array(new Int8Array(${removeTypedArrayPrefixFromInspectResult((0, util_1.inspect)(this.toInt8Array(), {
                    depth,
                    ...options,
                    maxArrayLength: util_1.inspect.defaultOptions.maxArrayLength,
                }))}))`;
            case bsonLibrary.Binary.VECTOR_TYPE.Float32:
                return `Binary.fromFloat32Array(new Float32Array(${removeTypedArrayPrefixFromInspectResult((0, util_1.inspect)(this.toFloat32Array(), {
                    depth,
                    ...options,
                    maxArrayLength: util_1.inspect.defaultOptions.maxArrayLength,
                }))}))`;
            case bsonLibrary.Binary.VECTOR_TYPE.PackedBit: {
                const paddingInfo = this.buffer[1] === 0 ? '' : `, ${this.buffer[1]}`;
                return `Binary.fromPackedBits(new Uint8Array(${removeTypedArrayPrefixFromInspectResult((0, util_1.inspect)(this.toPackedBits(), {
                    depth,
                    ...options,
                    maxArrayLength: util_1.inspect.defaultOptions.maxArrayLength,
                }))})${paddingInfo})`;
            }
            default:
                return binaryInspect.call(this, depth, options);
        }
    };
};
const makeBsonStringifiers = (bsonLibrary) => {
    const binaryVectorInspect = makeBinaryVectorInspect(bsonLibrary);
    const binaryInspect = makeClasslessInspect(bsonLibrary.Binary);
    return {
        ObjectId: makeClasslessInspect(bsonLibrary.ObjectId),
        ObjectID: makeClasslessInspect(bsonLibrary.ObjectId),
        DBRef: makeClasslessInspect(bsonLibrary.DBRef),
        MaxKey: makeClasslessInspect(bsonLibrary.MaxKey),
        MinKey: makeClasslessInspect(bsonLibrary.MinKey),
        Timestamp: makeClasslessInspect(bsonLibrary.Timestamp),
        BSONSymbol: makeClasslessInspect(bsonLibrary.BSONSymbol),
        Code: makeClasslessInspect(bsonLibrary.Code),
        Decimal128: makeClasslessInspect(bsonLibrary.Decimal128),
        Int32: makeClasslessInspect(bsonLibrary.Int32),
        Long: makeClasslessInspect(bsonLibrary.Long),
        Double: makeClasslessInspect(bsonLibrary.Double),
        BSONRegExp: makeClasslessInspect(bsonLibrary.BSONRegExp),
        Binary: function (...args) {
            const hexString = this.toString('hex');
            switch (this.sub_type) {
                case bsonLibrary.Binary.SUBTYPE_VECTOR:
                    return binaryVectorInspect.apply(this, args);
                case bsonLibrary.Binary.SUBTYPE_MD5:
                    return `MD5('${hexString}')`;
                case bsonLibrary.Binary.SUBTYPE_UUID:
                    if (hexString.length === 32) {
                        const asUUID = /^(.{8})(.{4})(.{4})(.{4})(.{12})$/
                            .exec(hexString)
                            .slice(1, 6)
                            .join('-');
                        return `UUID('${asUUID}')`;
                    }
                default:
                    return binaryInspect.apply(this, args);
            }
        },
    };
};
exports.makeBsonStringifiers = makeBsonStringifiers;
function makePrintableBson(bsonLibrary) {
    for (const [key, stringifier] of Object.entries((0, exports.makeBsonStringifiers)(bsonLibrary))) {
        if (!(key in bsonLibrary)) {
            continue;
        }
        const cls = bsonLibrary[key];
        for (const key of [inspectCustom, 'inspect']) {
            cls.prototype[key] = stringifier;
        }
    }
}
